Checkout: Strip products from url after checkout loads

sirbrillig opened the pull request:
#### Changes proposed in this Pull Request

There are specific URLs which will add products to the cart when loading checkout. For example, something like `/checkout/example.com/premium` will add the Premium plan to the cart and `/checkout/value_bundle/renew/10000001/example.com` will add a Premium plan renewal to the cart. However, keeping these URLs in the browser history has risks; if a user returns to this URL accidentally (for example, by using the browser's "Back" button) they would end up with the same products in their cart once again. This can result in duplicate purchases or unintended renewals.

This PR modifies checkout so that after the URL is parsed for products, it replaces the URL in the browser history with `/checkout/example.com`.

Some care is needed here due to the declarative nature of React. When the URL is modified, it changes the props passed in to the checkout components. If the URL is modified too soon, the products won't be added correctly to the cart. There may be other unintended effects.

Fixes https://github.com/Automattic/wp-calypso/issues/49368

#### Testing instructions

- Visit checkout with a URL like `/checkout/example.com/premium`
- Verify that the URL becomes `/checkout/example.com`
- Verify that the cart loads correctly with the product included.
- Repeat the above example with a renewal (it's easiest to use the "Renew now" button on the product page).

Consider other possible issues this may cause.

sirbrillig committed:
Strip products from url after checkout loads

sirbrillig committed:
Add no-site when siteSlug is not present

sirbrillig committed:
Correct renewal url example

nbloomf commented:
Urls are scary because we use them for several different things at once. Here are the scenarios I tested. I found one definite bug that affects both upgrades and renewals, and one less severe maybe-bug that only affects renewals, having to do with the X button.

**Entering checkout with a renewal from the purchase settings page:**
- Immediately click the back button (Result: ended up back on the purchase settings page as expected)
- Immediately refresh (Result: checkout reloaded with the correct products in the cart as expected)
- Immediately click the X (Result: ended up back on the purchase settings page as expected)
- Immediately refresh and then click the back button (Result: ended up on the purchase settings page as expected)
- Immediately refresh and then click the X (Result: ended up on the _plans_ page with the renewal in my cart. **This may be a bug.** I sort of expected to end up on the purchase settings page, although the semantics of "X" are less clear than those of "Back".)
- Complete step 2 (which puts a `#step2` fragment in the url) and click the back button (Result: stayed in checkout with step 2 in its editable state with my details still filled, as expected, or at least not unexpected)
- Complete step 2, click the back button, and refresh (Result: checkout reloads with my details not filled in, as expected)
- Complete step 2, click the back button, refresh, and click the X (Result: redirected to the plans page. I expected to end up on the purchase settings page; **this may be a bug**.)
- Complete step 2, click the back button, and then click the X. (Result: checkout reloads. I expected to go back to the purchase settings page. **I think this is a bug.** Clicking the X a second time redirects to the plans page.)
- Complete step 2 and refresh (Result: checkout reloads with the correct items in the cart and step 2 empty)
- Complete checkout and click the back button (Result: after checkout I am redirected back to the purchase settings page. On clicking "back", checkout briefly starts to load before redirecting to the plans page. This is probably the best possible outcome for in this situation.)

**Entering checkout with an upgrade from the plans page:**
- Immediately click the back button (Result: ended up back on the plans page with the upgrade in my cart, as expected)
- Immediately refresh (Result: checkout reloaded with the correct products in the cart as expected)
- Immediately click the X (Result: ended up back on the plans page with the upgrade in my cart, as expected)
- Immediately refresh and then click the back button (Result: ended up on the plans page with the upgrade in my cart, as expected)
- Immediately refresh and then click the X (Result: ended up back on the plans page with the upgrade in my cart, as expected)
- Complete step 2 and click the back button (Result: stayed in checkout with step 2 in its editable state with my details still filled, as expected)
- Complete step 2, click the back button, and refresh (Result: checkout reloads with my details not filled in, as expected)
- Complete step 2, click the back button, refresh, and click the X (Result: ended up on the plans page with the upgrade in my cart)
- Complete step 2, click the back button, and then click the X (Result: checkout reloads; I expected to go back to the plans page. **I think this is a bug.** Clicking the X a second time does go back to the plans page.)
- Complete step 2 and refresh (Result: checkout reloads with the correct cart and step 2 empty)
- Complete checkout and click the back button (Result: I purchased a premium plan upgrade and so ended up on the business plan upsell thank you page with url `/checkout/$site_url/offer-plan-upgrade/business/$receipt_id`. Clicking the back button here took me back to the plans page with an empty cart as expected.)

sirbrillig commented:
Another thought I just had is that coupons can be added to the cart via the query string, which is preserved in this implementation, and if we really want just a plain checkout URL we'll need to also remove that (but retain other query string params like `signup=1`).

sirbrillig commented:
> Immediately refresh and then click the X (Result: ended up on the plans page with the renewal in my cart. This may be a bug. I sort of expected to end up on the purchase settings page, although the semantics of "X" are less clear than those of "Back".)

This seems to be the behavior in production as well (please correct me if I'm wrong) so if it's a bug, I think it's unrelated to these changes.

> Complete step 2, click the back button, refresh, and click the X (Result: redirected to the plans page. I expected to end up on the purchase settings page; this may be a bug.)

I think this also happens in production. At least, it does for me.

> Complete step 2, click the back button, and then click the X. (Result: checkout reloads. I expected to go back to the purchase settings page. I think this is a bug. Clicking the X a second time redirects to the plans page.)

This also seems to happen in production when I tried it. I think the issue is that the "X" button uses `previousPath`, which comes from the `getPreviousPath()` Redux selector and probably can't tell certain checkout URLs apart. It looks like there was already some attempt to override the path detection for certain checkout routes. That logic could probably be improved but unless I'm missing something (always possible!) I don't think it's related to this PR.

https://github.com/Automattic/wp-calypso/blob/84fb7fcfd6c98086daa44ef996aeac151bf4f0b7/client/layout/masterbar/checkout.jsx#L19-L36

sirbrillig committed:
Remove coupon from query string

sirbrillig commented:
There's probably two things that could be done to deal with the "X" button issues:

1. Figure out how the previous page is stored in Redux and, for checkout, make sure it properly represents the page we actually came from.
2. Actively prevent the button from reloading checkout.

I've made a PR to handle number 2 in https://github.com/Automattic/wp-calypso/pull/49971

I'll take a look at number 1 shortly.

sirbrillig commented:
> Complete step 2, click the back button, refresh, and click the X (Result: redirected to the plans page. I expected to end up on the purchase settings page; this may be a bug.)

Wait, nevermind. I can somehow reproduce this now only in this branch. I must have been testing wrong before? I'm not sure. I'm going to look into this some more.

sirbrillig committed:
Prevent checkout masterbar close from reloading checkout

sirbrillig committed:
Move useStripProductsFromUrl to own file

sirbrillig committed:
Fix issue with querystring and hash not being added

sirbrillig committed:
Change checkout masterbar close to use redirect_to query string

sirbrillig commented:
Ok, I modified the "X" button behavior so that it will not be able to reload checkout. This could be a risk if there's another page whose route contains `/checkout/` that might be a legitimate "previous page" for the checkout masterbar close button. I think that the only routes with `/checkout/` are checkout itself, the thank-you pages, and the upsell pages, and none of those seem to me like a place I'd want the "X" button to return us to or a place where I'd expect the "X" to bring us to checkout so I think this is safe.

I then added code so that the "X" button will use the `redirect_to` query string parameter, if it's set in the URL. Even though that parameter is intended for _after_ checkout, I think it should be safe to use for this purpose as well. Notably, if you enter checkout from a renewal link on the subscriptions page, it will return you there when you click "X".

Hopefully that should solve all the issues you reported above.

sirbrillig committed:
Make sure stripped query string does not remove too many ampersands

sirbrillig committed:
Use cart action promises in useAddProductsFromUrl

Previously we were trying to guess that the cart actions were complete
by waiting for the cart status to change. However, if you apply a coupon
which is already applied, the cart status won't change, which will cause
this hook to be pending forever and make checkout never load.

Here we instead use the (relatively new) promises feature of cart
actions which allows us to actively wait for the actions to complete
which will happen even if the coupon already exists.

sirbrillig committed:
Clarify the second argument to useStripProductsFromUrl

sirbrillig committed:
Do not strip coupons after all

nbloomf commented:
I don't see the weird behavior with the X button anymore. Nice!
